testing/kokoro/system_tests.sh

```diff
@@ -44,25 +44,27 @@ export GOLANG_SAMPLES_E2E_TEST=""
 # allow files to be owned by a different user than our current uid.
 # Kokoro runs a double-nested container, and UIDs may not match.
 git config --global --add safe.directory $(pwd)
-GIT_CHANGES=$(git --no-pager diff --name-only main..HEAD)
+# Allow $GIT_CHANGES to be set in the env, enabling local testing of the change detection below.
+GIT_CHANGES=${GIT_CHANGES:-$(git --no-pager diff --name-only main..HEAD)}
 if [[ -z $GIT_CHANGES && $KOKORO_JOB_NAME != *"system-tests"* ]]; then
   echo "No diffs detected. This is unexpected - check above for additional error messages."
   exit 2
 fi
-SIGNIFICANT_CHANGES=$(git --no-pager diff --name-only main..HEAD | grep -Ev '(\.md$|^\.github)' || true )
+SIGNIFICANT_CHANGES=$(echo $GIT_CHANGES | grep -Ev '(\.md$|^\.github)' || true )
 # CHANGED_DIRS is the list of significant top-level directories that changed,
 # but weren't deleted by the current PR.
 # CHANGED_DIRS will be empty when run on main.
 CHANGED_DIRS=$(echo "$SIGNIFICANT_CHANGES" | tr ' ' '\n' | grep "/" | cut -d/ -f1 | sort -u | tr '\n' ' ' | xargs --no-run-if-empty ls -d 2>/dev/null || true)
+GO_CHANGED_PKGS=$(echo "$SIGNIFICANT_CHANGES" | tr ' ' '\n' | grep "/" | tr '\n' ' ' | xargs --no-run-if-empty dirname | xargs --no-run-if-empty ls -d 2>/dev/null || true)
```

Comment:
question: What is `xargs ls -d` meant to be doing? Some of these chained shellscripts seem like they are handling some undocumented edge cases.

Reply:
its ensuring that the directory still exists at HEAD (i.e. was not deleted). Tests dont run on deleted directories.
a better approach would be to use the output of `git diff --porcelain` to exclude deletions, but that can be done later as needed.